File: booster-commons/src/test/java/io/github/booster/commons/circuit/breaker/CircuitBreakerConfigTest.java
package io.github.booster.commons.circuit.breaker;

import io.github.booster.commons.metrics.MetricsRegistry;
import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.hamcrest.core.IsNull.notNullValue;
import static org.junit.jupiter.api.Assertions.assertThrows;

class CircuitBreakerConfigTest {

    @Test
    void shouldThrowException() {
        assertThrows(
                UnsupportedOperationException.class,
                () -> new CircuitBreakerConfig().getOption("test")
        );
        assertThrows(
                UnsupportedOperationException.class,
                () -> new CircuitBreakerConfig(Map.of()).getOption("test")
        );
        assertThrows(
                UnsupportedOperationException.class,
                () -> new CircuitBreakerConfig(Map.of("test", new CircuitBreakerSetting())).getOption("test")
        );
    }

    @Test
    void shouldCreateConfig() {
        assertThat(new CircuitBreakerConfig(), notNullValue());
        assertThat(new CircuitBreakerConfig(Map.of()), notNullValue());
        assertThat(new CircuitBreakerConfig(Map.of("test", new CircuitBreakerSetting())), notNullValue());
    }

    @Test
    void shouldNotCreateCircuitBreaker() {
        assertThat(new CircuitBreakerConfig().getOption("test").isDefined(), equalTo(false));
        assertThat(new CircuitBreakerConfig(Map.of()).getOption("test").isDefined(), equalTo(false));
        assertThat(
                new CircuitBreakerConfig(Map.of("abc", new CircuitBreakerSetting())).getOption("test").isDefined(),
                equalTo(false)
        );
    }

    @Test
    void shouldCreateCircuitBreaker() {
        assertThat(
                new CircuitBreakerConfig(Map.of("test", new CircuitBreakerSetting())).getOption("test").isDefined(),
                equalTo(true)
        );
    }

    @Test
    void shouldHandleRegistry() {
        CircuitBreakerConfig config = new CircuitBreakerConfig(Map.of("test", new CircuitBreakerSetting()));

        config.setMetricsRegistry(null);
        assertThat(
                config.getOption("test").isDefined(), equalTo(true)
        );
        assertThat(
                config.getOption("abc").isDefined(), equalTo(false)
        );

        config.setMetricsRegistry(new MetricsRegistry());
        assertThat(
                config.getOption("test").isDefined(), equalTo(true)
        );
        assertThat(
                config.getOption("abc").isDefined(), equalTo(false)
        );

        config.setMetricsRegistry(new MetricsRegistry(new SimpleMeterRegistry()));
        assertThat(
                config.getOption("test").isDefined(), equalTo(true)
        );
        assertThat(
                config.getOption("abc").isDefined(), equalTo(false)
        );
    }
}

File: booster-commons/src/test/kotlin/io/github/booster/commons/circuit/breaker/CircuitBreakerConfigTest.kt
package io.github.booster.commons.circuit.breaker

import io.github.booster.commons.metrics.MetricsRegistry
import io.micrometer.core.instrument.simple.SimpleMeterRegistry
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.notNullValue
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.Test

internal class CircuitBreakerConfigTest {
    @Test
    fun shouldCreateConfig() {
        assertThat(CircuitBreakerConfig(), notNullValue())
        assertThat(CircuitBreakerConfig(mapOf()), notNullValue())
        assertThat(CircuitBreakerConfig(mapOf(Pair("test", CircuitBreakerSetting()))), notNullValue())
    }

    @Test
    fun shouldNotCreateCircuitBreaker() {
        assertThat(CircuitBreakerConfig().tryGet("test").isDefined(), equalTo(false))
        assertThat(CircuitBreakerConfig(mapOf()).tryGet("test").isDefined(), equalTo(false))
        assertThat(
            CircuitBreakerConfig(mapOf(Pair("abc", CircuitBreakerSetting()))).tryGet("test").isDefined(),
            equalTo(false)
        )
    }

    @Test
    fun shouldCreateCircuitBreaker() {
        assertThat(
            CircuitBreakerConfig(mapOf(Pair("test", CircuitBreakerSetting()))).tryGet("test").isDefined(),
            equalTo(true)
        )
    }

    @Test
    fun shouldHandleRegistry() {
        val config = CircuitBreakerConfig(mapOf(Pair("test", CircuitBreakerSetting())))
        config.setMetricsRegistry(null)
        assertThat(
            config.tryGet("test").isDefined(), equalTo(true)
        )
        assertThat(
            config.tryGet("abc").isDefined(), equalTo(false)
        )
        config.setMetricsRegistry(MetricsRegistry())
        assertThat(
            config.tryGet("test").isDefined(), equalTo(true)
        )
        assertThat(
            config.tryGet("abc").isDefined(), equalTo(false)
        )
        config.setMetricsRegistry(MetricsRegistry(SimpleMeterRegistry()))
        assertThat(
            config.tryGet("test").isDefined(), equalTo(true)
        )
        assertThat(
            config.tryGet("abc").isDefined(), equalTo(false)
        )
    }
}
